Initial

doudou opened the pull request:
Depends on:
- [x] https://github.com/tidewise/drivers-nmea2000/pull/3
- [x] https://github.com/tidewise/tidewise.common-package_set/pull/12

doudou committed:
factor out the device dispatching code and implement CANTaskimplement CANTask and ActisenseTask

The project was originally to only do listening, and to avoid
dynamic dispatches to ports -- since messages on the NMEA2000 bus
are really low frequency. However, it turns out that the standard
NMEA2000 mechanisms to easily recognize to devices (instance numbers)
aren't supported by all devices, among which the fuel gauges we
have.

Since I don't want to shell out 100 eur for a programming device
each time someone feels like strongarming me, I've decided to implement
device enumeration on the bus, and allow to discriminate the devices
on their serial number. That's DeviceDispatcher

doudou committed:
add missing PGN reassembly in CANTask

doudou committed:
refactor to allow for continuous device enumeration

doudou committed:
mark can_in as needs_reliable_connection

doudou committed:
successfully query the product info of the next source once the current reply is received

doudou committed:
implement FluidLevelTask

doudou committed:
implement AISTask

doudou committed:
implement AirmarDepthSounderTask

doudou commented:
@henriquejsfj I think it would be good for you to have a read as well ... lots of stuff there, will raise useful questions.

doudou committed:
need to check query state and query probe message after processing incoming messages

CANTask is port triggered, so it's important that we process
the state change brought by a query answer (ProductInformation)
just after it was received.

doudou committed:
CANTask: add missing addPort and removePort calls

doudou committed:
be stricter on the check for m_received_product_info

Only set it when the product info we receive matches the last one
we sent

g-arjones commented on nmea2000.orogen:
```diff
@@ -9,48 +9,110 @@ using_task_library 'iodrivers_base'
+    property 'enumeration_ack_timeout', '/base/Time'
+
+    # Dispatched messages as set up by the 'devices' property
+    dynamic_output_port(/\w+/, '/nmea2000/Message')
```
Didn't know about `dynamic_output_port`... Is this new? What's the regex for?

doudou replied:
Not new. It's to tell Syskit that the component has the ability to create new ports. The regexp is to identify naming schemes when there are some (not the case here).

g-arjones replied:
Meaning it will break on instantiation if the name of a dynamically created port does not follow the scheme?

doudou replied:
- Syskit will refuse to generate networks with dynamic ports if there isn't a matching dynamic port declaration on the orogen file
- Syskit obviously verifies that the dynamic ports it expects have been created by the component

However, there is no help on the C++ side, that is the component implementer could very well create ports that do not match the declaration. I've been thinking for a while to add a `as: NAME` option to `dynamic_*_port` which would generate code to manage the ports for the component implementer, but never actually did it.

g-arjones commented on tasks/DeviceDispatcher.hpp:
```diff
@@ -0,0 +1,93 @@
+            uint8_t bus_id = 0;
+            uint32_t pgn;
+
+            RTT::OutputPort<Message>* port = nullptr;
```
I think you would make addPorts/removePorts safer if you used a smart pointer here

doudou replied:
I don't see how. Can you expand ?

g-arjones replied:
First, you don't have to worry about deleting the dynamically allocated ports (which I just noticed actually don't do in `removePorts`...). You could also use unique_ptr's custom deleter to not even have to worry about calling removePorts before addPorts.. Just replace them and let the deleter do the removing...

doudou replied:
Deletion and removePorts have to work hand-in-hand, or things will crash if the unique_ptr is cleared before removePorts is called anyways. So, to me, this simply removes one problem to create others.

g-arjones replied:
Maybe.. But the current implementation leaks the ports if they are changed doesn't it?

doudou replied:
Yes. So ?

[The example in the README](https://github.com/rock-core/drivers-orogen-iodrivers_base) crashes on error, but is still using `unique_ptr` and RAII. Just pointing out that "not leaking" is not "perfect code", and that therefore "leaking" does not mandate "switch to smart pointers".

doudou replied:
Anyways, fixed the very broken logic of removePort, and cleaned addPort a little.

g-arjones replied:
I don't know where the "perfect code" argument came from, we both know there is no such thing. 

And "leaking" does not mandate "smart pointers" but smart pointers definitely makes leaking less likely. To be honest, other than a slight performance hit, I don't see any downsides to using smart pointers, therefore I believe they should be used whenever possible in modern C++ code.

Anyway, the code looks fine now...

doudou replied:
> I believe they should be used whenever possible in modern C++ code.

I don't agree. I think they make things harder to understand when misused. `shared_ptr` in particular can be very easily misused, leading to code that feels like you have a GC, but since you don't things end up being very messy.

At the end of the day, they are IMO great for cases where lifetimes are clear-cut. A.k.a. there are no or very little interaction between the lifetime of an object and either other objects' operations, or the lifetime of other objects. As soon as their are more complex interactions, I use them very sparingly because they IMO make things very difficult. The iodrivers_base readme is a very good example to me of how things can go wrong *because* these effects were hidden from view, and that was with 10 lines of code.

g-arjones replied:
There's definitely an argument to that. However, I don't think the iodrivers_base thing is a good example... That was misuse of RAII, not smart pointers (the same would happen if `Driver` was a stack allocated object).

doudou committed:
fix lifecycle management of the ports in DeviceDispatcher

- avoid creating any ports if the configuration is invalid
  - add test to check whether a port is already used by the task's
    own interface
- avoid calling task.addPort for an already registered port
- fix ports being leaked on removePort
  - make sure removePort is compatible with port sharing
